Kernel/Arch/x86_64/Processor.h

```diff
@@ -40,36 +40,25 @@ struct ProcessorMessageEntry;
 #define MSR_IA32_EFER 0xc0000080
 #define MSR_IA32_PAT 0x277
 
-// FIXME: Find a better place for these
-extern "C" void thread_context_first_enter(void);
-extern "C" void exit_kernel_thread(void);
-extern "C" void do_assume_context(Thread* thread, u32 flags);
-
-struct [[gnu::aligned(64), gnu::packed]] FPUState {
-    SIMD::LegacyRegion legacy_region;
-    SIMD::Header xsave_header;
-
-    // FIXME: This should be dynamically allocated! For now, we only save the `YMM` registers here,
-    // so this will do for now. The size of the area is queried via CPUID(EAX=0dh, ECX=2):EAX.
-    // https://www.intel.com/content/dam/develop/external/us/en/documents/36945
-    u8 ext_save_area[256];
-};
-
 enum class InterruptsState;
 class Processor;
+
+template<typename ProcessorT>
+class ProcessorBase;
+
 // Note: We only support 64 processors at most at the moment,
 // so allocate 64 slots of inline capacity in the container.
 
 constexpr size_t MAX_CPU_COUNT = 64;
 using ProcessorContainer = Array<Processor*, MAX_CPU_COUNT>;
 
-class Processor {
-    friend class ProcessorInfo;
+extern "C" void context_first_init(Thread* from_thread, Thread* to_thread, [[maybe_unused]] TrapFrame* trap);
 
-    AK_MAKE_NONCOPYABLE(Processor);
-    AK_MAKE_NONMOVABLE(Processor);
-
-    Processor* m_self;
+// If this fails to compile because ProcessorBase was not found, you are including this header directly.
+// Include Arch/Processor.h instead.
+class Processor final : public ProcessorBase<Processor> {
+    friend class ProcessorInfo;
+    friend class ProcessorBase<Processor>;
```

Comment:
I really skeptical of this friendship. Why do we need to friend our base class? If we need to do that, just mark the relevant private methods as protected